castai/resource_workload_scaling_policy.go

```diff
@@ -0,0 +1,342 @@
+				ValidateDiagFunc: validation.ToDiagFunc(validation.StringInSlice([]string{"READ_ONLY", "MANAGED"}, false)),
+			},
+			"cpu": {
+				Type:     schema.TypeList,
+				Required: true,
+				MaxItems: 1,
+				Elem:     resourceSchema("QUANTILE", 0),
+			},
+			"memory": {
+				Type:     schema.TypeList,
+				Required: true,
+				MaxItems: 1,
+				Elem:     resourceSchema("MAX", 0.1),
+			},
+		},
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(15 * time.Second),
+			Read:   schema.DefaultTimeout(15 * time.Second),
+			Update: schema.DefaultTimeout(15 * time.Second),
+			Delete: schema.DefaultTimeout(15 * time.Second),
+		},
+	}
+}
+
+func resourceSchema(function string, overhead float64) *schema.Resource {
+	return &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"function": {
+				Type:             schema.TypeString,
+				Optional:         true,
+				Description:      "The function used to calculate the resource recommendation. Supported values: `QUANTILE`, `MAX`",
+				Default:          function,
+				ValidateDiagFunc: validation.ToDiagFunc(validation.StringInSlice([]string{"QUANTILE", "MAX"}, false)),
+			},
+			"args": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Description: "The arguments for the function - i.e. for `QUANTILE` this should be a [0, 1] float. " +
+					"`MAX` doesn't accept any args",
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"overhead": {
+				Type:             schema.TypeFloat,
+				Optional:         true,
+				Description:      "Overhead for the recommendation, e.g. `0.1` will result in 10% higher recommendation",
+				Default:          overhead,
+				ValidateDiagFunc: validation.ToDiagFunc(validation.FloatBetween(0, 1)),
+			},
+			"apply_threshold": {
+				Type:     schema.TypeFloat,
+				Optional: true,
+				Description: "The threshold of when to apply the recommendation. Recommendation will be applied when " +
+					"diff of current requests and new recommendation is greater than set value",
+				Default:          0.1,
+				ValidateDiagFunc: validation.ToDiagFunc(validation.FloatBetween(0.01, 1)),
+			},
+		},
+	}
+}
+
+func resourceWorkloadScalingPolicyCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*ProviderConfig).api
+
+	clusterID := d.Get(FieldClusterID).(string)
+	req := sdk.WorkloadOptimizationAPICreateWorkloadScalingPolicyJSONRequestBody{
+		Name:      d.Get("name").(string),
+		ApplyType: sdk.WorkloadoptimizationV1ApplyType(d.Get("apply_type").(string)),
+		RecommendationPolicies: sdk.WorkloadoptimizationV1RecommendationPolicies{
+			ManagementOption: sdk.WorkloadoptimizationV1ManagementOption(d.Get("management_option").(string)),
+		},
+	}
+
+	if v, ok := d.GetOk("cpu"); ok {
+		req.RecommendationPolicies.Cpu = toResourcePolicies(v.([]interface{})[0].(map[string]interface{}))
```

Comment:
Similar to what I mentioned in another comment. Could you please rename the function? Something that narrows the context.